File: examples/typescript/x402-request.ts
/**
 * Runnable example: x402 Request (direct settlement, v2 wire format)
 *
 * Usage:
 *   PAYSKILL_KEY=0x... npx tsx x402-request.ts
 *
 * This demo starts a local 402 test server, then uses PayClient.request()
 * to auto-pay and retry.
 */

import { createServer } from "node:http";
import { PayClient } from "@pay-skill/sdk";

const API_URL = "https://testnet.pay-skill.com/api/v1";

async function main() {
  const key = process.env.PAYSKILL_KEY;
  if (!key) throw new Error("Set PAYSKILL_KEY env var");

  // Fetch contract addresses — never hardcode these
  const contracts = await fetch(`${API_URL}/contracts`).then((r) => r.json());

  // Start a local test server that returns v2 402
  const server = createServer((req, res) => {
    const sig = req.headers["payment-signature"];
    if (sig && typeof sig === "string" && sig.length > 0) {
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ content: "premium data", paid: true }));
    } else {
      // v2 PaymentRequired
      const paymentRequired = {
        x402Version: 2,
        resource: { url: "/data", mimeType: "application/json" },
        accepts: [{
          scheme: "exact",
          network: `eip155:${contracts.chain_id}`,
          amount: "1000000",
          asset: contracts.usdc,
          payTo: "0x000000000000000000000000000000000000dEaD",
          maxTimeoutSeconds: 60,
          extra: { settlement: "direct" },
        }],
        extensions: {},
      };
      const encoded = Buffer.from(JSON.stringify(paymentRequired)).toString("base64");
      res.writeHead(402, {
        "Content-Type": "application/json",
        "PAYMENT-REQUIRED": encoded,
      });
      res.end(JSON.stringify(paymentRequired));
    }
  });

  await new Promise<void>((r) => server.listen(0, "127.0.0.1", r));
  const port = (server.address() as { port: number }).port;
  console.log(`Test server on http://127.0.0.1:${port}`);

  try {
    const client = new PayClient({
      apiUrl: API_URL,
      privateKey: key,
      chainId: contracts.chain_id,
      routerAddress: contracts.router,
    });

    console.log("Requesting (will auto-pay on 402)...");
    const resp = await client.request(`http://127.0.0.1:${port}/data`);
    const body = await resp.json();
    console.log("Status:", resp.status);
    console.log("Body:", body);
  } finally {
    server.close();
  }
}

main().catch(console.error);

File: examples/python/x402_request.py
"""
Runnable example: x402 Request (v2 wire format)

Usage:
    PAYSKILL_KEY=0x... python x402_request.py

Starts a local 402 test server, then auto-pays via PayClient.request().
"""

import os
import json
import base64
import threading
from http.server import HTTPServer, BaseHTTPRequestHandler
from payskill import PayClient

API_URL = "https://testnet.pay-skill.com/api/v1"
ROUTER = "0x24F26eCb1f46451994c59585817e87896749935D"


class Handler(BaseHTTPRequestHandler):
    def do_GET(self) -> None:
        sig = self.headers.get("Payment-Signature", "")
        if sig:
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({"content": "premium data", "paid": True}).encode())
        else:
            # v2 PaymentRequired
            payment_required = {
                "x402Version": 2,
                "resource": {"url": "/data", "mimeType": "application/json"},
                "accepts": [{
                    "scheme": "exact",
                    "network": "eip155:84532",
                    "amount": "1000000",
                    "asset": "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
                    "payTo": "0x000000000000000000000000000000000000dEaD",
                    "maxTimeoutSeconds": 60,
                    "extra": {
                        "name": "USDC",
                        "version": "2",
                        "facilitator": "https://testnet.pay-skill.com/x402",
                        "settlement": "direct",
                    },
                }],
                "extensions": {},
            }
            encoded = base64.b64encode(json.dumps(payment_required).encode()).decode()
            self.send_response(402)
            self.send_header("Content-Type", "application/json")
            self.send_header("PAYMENT-REQUIRED", encoded)
            self.end_headers()
            self.wfile.write(json.dumps(payment_required).encode())

    def log_message(self, fmt: str, *args: object) -> None:
        pass  # suppress logs


def main() -> None:
    key = os.environ.get("PAYSKILL_KEY")
    if not key:
        raise RuntimeError("Set PAYSKILL_KEY env var")

    server = HTTPServer(("127.0.0.1", 0), Handler)
    port = server.server_address[1]
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    print(f"Test server on http://127.0.0.1:{port}")

    try:
        client = PayClient(
            api_url=API_URL, signer="raw", private_key=key,
            chain_id=84532, router_address=ROUTER,
        )

        print("Requesting (will auto-pay on 402)...")
        resp = client.request(f"http://127.0.0.1:{port}/data")
        print(f"Status: {resp.status_code}")
        print(f"Body: {resp.json()}")
    finally:
        server.shutdown()


if __name__ == "__main__":
    main()
